dns/dns-loadbalancer-rr.py

```diff
@@ -0,0 +1,80 @@
+#!/usr/bin/python3
+
+# This proof-of-concept is a variant of dns-loadbalancer.py. Whereas the
+# latter returns one random Scylla node in response to every request, the
+# implementation in this file returns the *entire* list of live nodes, shifted
+# cyclically by a random amount. This technique allows clients which can make
+# use of the entire list to use it - while clients who can't and use the entire
+# list and use the first one will still get a random node, as happens in
+# dns-loadbalancer.py. This technique is known as "Round-robin DNS" - see
+#    https://en.wikipedia.org/wiki/Round-robin_DNS
+
+import dnslib.server
+import dnslib
+import random
+import _thread
+import urllib.request
+import time
+
+# The list of live nodes, all of them supposedly answering HTTP requests on
+# alternator_port. All of these nodes will be returned, shifted by a random
+# amount, from every DNS request.
+# This list starts with one or more known nodes, but then the
+# livenodes_update() thread periodically replaces this list by an up-to-date
+# list retrieved from makeing a "/localnodes" requests to one of these nodes.
+livenodes = ['127.0.0.1']
+alternator_port = 8000
+def livenodes_update():
+    global alternator_port
+    global livenodes
+    while True:
+        # Contact one of the already known nodes by random, to fetch a new
+        # list of known nodes.
+        # TODO: We could reuse the HTTP connection (and prefer to make the
+        # request to the same node again, and not a random node).
+        ip = random.choice(livenodes)
+        url = 'http://{}:{}/localnodes'.format(ip, alternator_port)
+        print('updating livenodes from {}'.format(url))
+        try:
+            nodes = urllib.request.urlopen(url, None, 1.0).read().decode('ascii')
+            a = [x.strip('"').rstrip('"') for x in nodes.strip('[').rstrip(']').split(',')]
+            # If we're successful, replace livenodes by the new list
+            livenodes = a
+            print(livenodes)
+        except:
+            # Contacting this ip was unsuccessful, we could remove remove it
+            # from the list of live nodes, but tais is not a good idea if
+            # all nodes are temporarily down. In any case, when we do reach
+            # a live node, we'll replace the entire list.
+            pass
+        time.sleep(1)
+_thread.start_new_thread(livenodes_update,())
+
+def random_shift(l):
+    shift = random.randrange(len(l))
+    return l[shift::] + l[:shift:]
```

Comment:
This one should be quicker
```suggestion
def random_shift(l):
    shift = random.randrange(len(l))
    length = len(l)
    for i in range(length):
        yield l[(shift + i) % length]
```

Reply:
Your generator-based approach doesn't allocate memory, but I can't imagine it is quicker, I am guessing (without measuring, I admit) that it's actually slower, because instead of allocating the result and filling it, it uses a lot of function calls to implement the generator.

Maybe something like this will also work (using the new Python 3.5 "yield from"):
```python
def random_shift(l):
     shift = random.randrange(len(l))
     yield from l[shift::]
     yield from l[:shift:]
```

It's shorter and prettier than your version, but still has quite a lot of function calls so fast it probably isn't.

I was surprised that Python doesn't have a builtin function for such cyclic shift - it has a function for everything else :-) Maybe it has and I'm just didn't find.

Reply:
`yield from l[shift::]` will also cause list allocation, anyways, python and performance is like oil and water. I don't mind merging as is.

Reply:
You're right - I thought `l[shift::]` is a generator, but turns out it isn't.
I still believe that performance-wise, `l[shift::] + l[:shift:]` is fastest because although it does do allocations and copies, all of this happens in C code, and not interpreted code like in your version. Avoiding allocation is good when the list is huge - e.g., a billion node - but obviously we'll never have this.